### PR TITLE
KTOR 1344 - Unified pre-typed routes

### PR DESCRIPTION
**Pull request by hfhbd**
**Subsystem**
Server - Routing

**Motivation**
Currently, only `post` has a typed overload. [KTOR-1344](https://youtrack.jetbrains.com/issue/KTOR-1344)

**Solution**
Added a typed overload to `put` and `patch` as well

### CODE REVIEW

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RoutingBuilder.kt [hunk 1]
```diff
@@ -118,11 +118,15 @@ public fun Route.post(path: String, body: PipelineInterceptor<Unit, ApplicationC
 @ContextDsl
 @JvmName("postTyped")
 public inline fun <reified R : Any> Route.post(
-    path: String,
+    path: String? = null,
     crossinline body: suspend PipelineContext<Unit, ApplicationCall>.(R) -> Unit
 ): Route {
-    return route(path, HttpMethod.Post) {
-        handle {
+    return if(path != null) {
```
Comment (cy6erGn0m) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RoutingBuilder.kt:124]: The problem here is that the function is inline so introducing `if` here you duplicate the body bytecode and duplicate user's bytecode everywhere regardless of whether they need the `path` or not.

**Review by cy6erGn0m — CHANGES_REQUESTED** [review]
LGTM except for bytecode duplication

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RoutingBuilder.kt (hunk 1 above)
Comment (hfhbd) [reply ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RoutingBuilder.kt:124]: So you need multiple overloads? One with path, one without?

**Comment by cy6erGn0m** [comment]
@rsinukov This generally looks fine but I feel these function may/will clash with locations functions so I wonder if we can somehow workaround it? The functions contributed seems useful but I don't know how do we avoid ambiguity

**Comment by rsinukov** [comment]
@cy6erGn0m it looks ok for me. Locations builders and these ones are in different packages, so users can always specify what function they want to use. We already have this for `post` and we never had any complaints about it.

**Review by rsinukov — CHANGES_REQUESTED** [review]
Please run `./gradlew apiDump` and change the target branch to 1.6.0

**Comment by hfhbd** [comment]
@rsinukov Changed. apiDump did not update my changes, because all functions are `inline reified`.

**Review by rsinukov — APPROVED** [review]

**Review by cy6erGn0m — APPROVED** [review]

**Comment by cy6erGn0m** [comment]
@hfhbd Ok, let's proceed with merge. Could you please rebase it and skip irrelevant commits? Notice the base branch 1.6.0-eap

**Comment by hfhbd** [comment]
Dont merge, one second

**Comment by hfhbd** [comment]
Okay, I don't care about the username
@cy6erGn0m rebased

**Pull request merged**